### PR TITLE
Federated Clients: implement LiKafkaFederatedConsumerImpl#assign() and assignment()

### PR DESCRIPTION
**Pull request by jonlee2**
Also removed an unused class from LiKafkaFederatedProducerImpl.java.

### CODE REVIEW

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 1]
```diff
@@ -349,4 +365,27 @@ public void close(Duration timeout) {
   public void wakeup() {
     throw new UnsupportedOperationException("Not implemented yet");
   }
+
+  public LiKafkaConsumer<K, V> getPerClusterConsumer(ClusterDescriptor cluster) {
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: does this need to be public ? we're trying to "fake" a simple client

Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: This is used for unit tests. I can add a comment saying that it is intended for testing only. Would that be acceptable?

Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Per our discussion, changed it to package private. Also added a comment.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 2]
```diff
@@ -137,7 +144,23 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener callb
...
+    Map<TopicPartition, ClusterDescriptor> topicPartitionToClusterMap =
+        _mdsClient.getClustersForTopicPartitions(_clientId, partitions, _mdsRequestTimeoutMs);
+
+    // Reverse the map so that we can have per-cluster topic partition sets.
+    Map<ClusterDescriptor, Set<TopicPartition>> clusterToTopicPartitionsMap = new HashMap<>();
+    for (Map.Entry<TopicPartition, ClusterDescriptor> entry : topicPartitionToClusterMap.entrySet()) {
+      TopicPartition topicPartition = entry.getKey();
+      ClusterDescriptor cluster = entry.getValue();
+      Set<TopicPartition> newTopicPartitionSet =
+          clusterToTopicPartitionsMap.getOrDefault(cluster, new HashSet<TopicPartition>());
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: clusterToTopicPartitionsMap.computeIfAbsent(cluster, {} -> new ...).add(topicPartition)

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 3]
```diff
@@ -349,4 +372,28 @@ public void close(Duration timeout) {
...
     throw new UnsupportedOperationException("Not implemented yet");
   }
+
+  // Intended for testing only
+  LiKafkaConsumer<K, V> getPerClusterConsumer(ClusterDescriptor cluster) {
+    return _consumers.get(cluster);
+  }
+
+  private LiKafkaConsumer<K, V> getOrCreateConsumerForTopic(String topic) {
+    return getOrCreatePerClusterConsumer(_mdsClient.getClusterForTopic(_clientId, topic, _mdsRequestTimeoutMs));
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: i think you should handle clusterForTopic() returning null

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 2 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Neat! Was looking for this capability.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 3 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Added a null check to getOrCreatePerClusterConsumer().

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 4]
```diff
@@ -137,7 +144,30 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener callb
 
   @Override
   public void assign(Collection<TopicPartition> partitions) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (partitions == null) {
+      throw new IllegalArgumentException("Topic partition collection to assign to cannot be null");
+    }
+
+    // If partitions are empty, it should be treated the same as unsubscribe().
+    if (partitions.isEmpty()) {
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: update the comment to reflect that this is the vanilla client behaviour (as its a weird one)

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 5]
```diff
@@ -137,7 +144,30 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener callb
...
+      unsubscribe();
+      return;
+    }
+
+    Map<TopicPartition, ClusterDescriptor> topicPartitionToClusterMap =
+        _mdsClient.getClustersForTopicPartitions(_clientId, partitions, _mdsRequestTimeoutMs);
+
+    // Reverse the map so that we can have per-cluster topic partition sets.
+    Map<ClusterDescriptor, Set<TopicPartition>> clusterToTopicPartitionsMap = new HashMap<>();
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: what about TopicPartitions for which no cluster was found?

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 4 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 5 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Argh.. I initially assumed that Mario would filter them out, but now that we decided to throw an error for nonexistent topics/partitions, we should check.. My bad.

**Pull request merged**